File: programs/codechef/july14_long/sgarden.cpp
#include <cstdio>
#include <cstring>
typedef long long LL;
using namespace std;
int mod = 1000000007;

int gcd(LL a, LL b) { return !a || !b ? a+b : gcd(b,a%b); }

int main()
{
	int t; scanf("%d",&t);
	while(t--)
	{
		int n; scanf("%d",&n);
		int ar[n], done[n];
		LL lcm=1;
		memset(done,0,sizeof(int)*n);
		for(int a=0;a<n;a++) { scanf("%d",ar+a); ar[a]--; }
		for(int a=0;a<n;a++)
		{
			if(done[a]) continue;
			int len=0,orig=a;
			do
			{
				len++;
				done[orig] = 1;
				orig = ar[orig];
			}while(orig!=a);
			int g = gcd(lcm,len);
			lcm = (lcm*len) / g;
			lcm %= mod;
			//printf("length: %d\n",len);
		}
		printf("%d\n",lcm);
	}
	return 0;
}

File: contests/codechef/july14_long/sgarden.py
def gcd(a,b):
	if a == 0 or b == 0: return a+b
	return gcd(b,a%b)

for t in range(input()):
	n = input()
	ar = [0] * n
	done = [0] * n
	s = map(int,raw_input().split())
	ar = [i-1 for i in s]
	lcm = 1
	for a in range(n):
		if done[a]: continue
		len = 1
		done[a] = 1
		orig = ar[a]
		while (orig!=a):
			len += 1
			done[orig] = 1
			orig = ar[orig]
		lcm = (lcm*len)/gcd(lcm,len);
	print(lcm%(10**9+7))
